crates/engine/local/src/miner.rs

```diff
@@ -58,3 +73,175 @@ impl Future for MiningMode {
+                _ = &mut self.mode => {
+                    if let Err(e) = self.advance().await {
+                        error!(target: "engine::local", "Error advancing the chain: {:?}", e);
+                    }
+                }
+                // send FCU once in a while
+                _ = fcu_interval.tick() => {
+                    if let Err(e) = self.update_forkchoice_state().await {
+                        error!(target: "engine::local", "Error updating fork choice: {:?}", e);
+                    }
+                }
+            }
+        }
+    }
+
+    /// Returns current forkchoice state.
+    fn forkchoice_state(&self) -> ForkchoiceState {
+        ForkchoiceState {
+            head_block_hash: *self.last_block_hashes.last().expect("at least 1 block exists"),
+            safe_block_hash: *self
+                .last_block_hashes
+                .get(self.last_block_hashes.len().saturating_sub(32))
+                .expect("at least 1 block exists"),
+            finalized_block_hash: *self
+                .last_block_hashes
+                .get(self.last_block_hashes.len().saturating_sub(64))
+                .expect("at least 1 block exists"),
+        }
+    }
+
+    /// Sends a FCU to the engine.
+    async fn update_forkchoice_state(&self) -> eyre::Result<()> {
+        let (tx, rx) = oneshot::channel();
+        self.to_engine.send(BeaconEngineMessage::ForkchoiceUpdated {
+            state: self.forkchoice_state(),
+            payload_attrs: None,
+            tx,
+        })?;
+
+        let res = rx.await??;
+        if !res.forkchoice_status().is_valid() {
+            eyre::bail!("Invalid fork choice update")
+        }
+
+        Ok(())
+    }
+
+    /// Generates payload attributes for a new block, passes them to FCU and inserts built payload
+    /// through newPayload.
+    async fn advance(&mut self) -> eyre::Result<()> {
+        let timestamp = std::cmp::max(
+            self.last_timestamp + 1,
+            std::time::SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("cannot be earlier than UNIX_EPOCH")
+                .as_secs(),
+        );
+
+        let (tx, rx) = oneshot::channel();
+        self.to_engine.send(BeaconEngineMessage::ForkchoiceUpdated {
+            state: self.forkchoice_state(),
+            payload_attrs: Some(self.payload_attributes_builder.build(timestamp)),
+            tx,
+        })?;
+
+        let res = rx.await??.await?;
+        if !res.payload_status.is_valid() {
+            eyre::bail!("Invalid payload status")
+        }
+
+        let payload_id = res.payload_id.ok_or_eyre("No payload id")?;
+
+        // wait for some time to let the payload be built
+        tokio::time::sleep(Duration::from_millis(200)).await;
+
+        let Some(Ok(payload)) = self.payload_builder.best_payload(payload_id).await else {
+            eyre::bail!("No payload")
+        };
```

Comment:
I think we want to call resolve here:

https://github.com/paradigmxyz/reth/blob/b6371018944e8cba6f5b4c56e76ff211b9e820b1/crates/payload/builder/src/service.rs#L44-L53

we likely want to change this as well

https://github.com/paradigmxyz/reth/blob/b6371018944e8cba6f5b4c56e76ff211b9e820b1/crates/payload/builder/src/traits.rs#L21-L25

so we we can tell the future whether we want to want to wait for the in progress payload or resolve it as fast as possible, but we can tackle this separately

Reply:
but we can do this in a followup